group-coordinator/src/main/java/org/apache/kafka/coordinator/group/OffsetMetadataManager.java

```diff
@@ -653,7 +676,17 @@ private void removeOffset(
         if (partitionOffsets.isEmpty())
             topicOffsets.remove(topic);
 
-        if (topicOffsets.isEmpty())
+        if (topicOffsets.isEmpty()) {
             offsetsByGroup.remove(groupId);
+            // A group in offset metadata manager should always exist in the group metadata manager.
+            // In the case the write operation fails, the generic group will remain Dead.
+            // This is okay since it follows the old coordinator behavior; delete expired offsets and
+            // transition group to Dead even if append fails. One caveat is that groups that just transitioned
+            // to Dead will be deleted in the next expiration cycle.
+            Group group = groupMetadataManager.group(groupId);
+            if (group.isEmpty()) {
+                group.transitionToDead();
```

Comment:
I actually wonder if we need this after all. If the deletion fails, the group will be restored at its previous state when the operation is reverted in the timeline data structure.